### DISCUSSION TITLE
NaN appears to be broken in V2

### DISCUSSION THREAD

**Discussion by hvmptydvmpty** [pydantic/pydantic#7973] [Question]
First of all thanks a lot for a great library! Now the question. I went and checked this in all major version of Pydantic between 2.0 and 2.4 and can't get _NaN_ to work. Initially this transpired from unit tests that worked fine in V1 but ground to a halt in V2. Also, an interesting note is that `parse_raw` works as expected and produces _NaN_.

```python
>>> import json, math, pydantic, typing
>>> class cls(pydantic.BaseModel):
    x: typing.Annotated[float, pydantic.Field(allow_inf_nan=True)]  # my original code was typing.List[typing.List[typing.Any]]

>>> json.dumps({"x": math.nan})
'{"x": NaN}'
>>> cls.model_validate_json(_)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File ".../lib/python3.9/site-packages/pydantic/main.py", line 530, in model_validate_json
    return cls.__pydantic_validator__.validate_json(json_data, strict=strict, context=context)
pydantic_core._pydantic_core.ValidationError: 1 validation error for cls
  Invalid JSON: expected value at line 1 column 7 [type=json_invalid, input_value='{"x": NaN}', input_type=str]
    For further information visit https://errors.pydantic.dev/2.4/v/json_invalid
```
No matter which combination of field definitions I try, annotated or not, parsing of NaN refuses to work. I also went and checked some examples of how it's done using `pydantic_core.core_schema.float_schema` and that doesn't work for me either.
Also tried on two different machines, macOS and Linux.

I can't believe this has gone unnoticed for four major versions so I must be doing it wrong but what exactly I can't find. Thank you for the help!

**Comment by samuelcolvin** [comment]
> First of all thanks a lot for a great library! 

Thanks a lot!

See #7007 for details.
